h3/src/quic.rs

```diff
@@ -29,16 +30,59 @@ impl<'a, E: Error + 'a> From<E> for Box<dyn Error + 'a> {
     }
 }
 
+/// Types of erros that *specifically* arises when sending a datagram.
+#[derive(Debug)]
+pub enum SendDatagramError {
+    /// Datagrams are not supported by the peer
+    UnsupportedByPeer,
+    /// Datagrams are locally disabled
+    Disabled,
+    /// The datagram was too large to be sent.
+    TooLarge,
+    /// Network error
+    ConnectionLost(Box<dyn Error>),
+}
+
+impl fmt::Display for SendDatagramError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            SendDatagramError::UnsupportedByPeer => write!(f, "datagrams not supported by peer"),
+            SendDatagramError::Disabled => write!(f, "datagram support disabled"),
+            SendDatagramError::TooLarge => write!(f, "datagram too large"),
+            SendDatagramError::ConnectionLost(_) => write!(f, "connection lost"),
+        }
+    }
+}
+
+impl std::error::Error for SendDatagramError {}
+
+impl Error for SendDatagramError {
+    fn is_timeout(&self) -> bool {
+        false
+    }
+
+    fn err_code(&self) -> Option<u64> {
+        match self {
+            Self::ConnectionLost(err) => err.err_code(),
+            _ => None,
+        }
+    }
+}
+
 /// Trait representing a QUIC connection.
-pub trait Connection<B: Buf> {
+pub trait Connection {
```

Comment:
So far, we've allowed implementations/users to specify a different `Buf` type than `Bytes`, in case they want to make better use of vectored writes, or just less copies in general. Seems like a good property to keep.

Reply:
If I've understood it correctly:

You want to allow

```rust
impl Connection<Vec<u8>> for MyConn { ... }
impl Connection<Bytes> for MyConn { ... }
impl Connection<Vec<Bytes>> for MyConn { ... }
```

Rather than a blanketed impl for *all* Buf types like quinn and s2n does?

In that case, perhaps there is a better way to allow this which doesn't constrain the user in what data they can send, or doesn't leak the underlying buffering type all the way up the the user API.

Reply:
Sort of, yea. More like in hyper and h2, they're just `Connection<Io, BufType>`, and whatever kind of the buf the user has, we can work with it.